src/Datadog.Trace.ClrProfiler.Managed.Loader/Startup.cs

```diff
@@ -58,10 +192,75 @@ private static string ReadEnvironmentVariable(string key)
             }
 
             return null;
         }
+
+        private void ResolveManagedProfilerDirectory()
+        {
+            Assembly objectAssembly = (new object()).GetType().Assembly;
+            bool isCoreFx = (objectAssembly?.FullName?.StartsWith("System.Private.CoreLib") == true);
+
+            string frameworkBasedSubdir;
+            if (isCoreFx)
+            {
+                frameworkBasedSubdir = "netstandard2.0";
+
+                // Old versions of .NET core report a major version of 4
+                Version clrVersion = Environment.Version;
+                if ((clrVersion.Major == 3 && clrVersion.Minor >= 1) || clrVersion.Major >= 5)
+                {
+                    frameworkBasedSubdir = "netcoreapp3.1";
+                }
+            }
+            else
+            {
+                // We currently build two assemblies targeting .NET Framework.
+                // If we're running on the .NET Framework, load the highest-compatible assembly
+                string corlibFileVersionString = ((AssemblyFileVersionAttribute)objectAssembly.GetCustomAttribute(typeof(AssemblyFileVersionAttribute))).Version;
+                string corlib461FileVersionString = "4.6.1055.0";
+
+                // This will throw an exception if the version number does not match the expected 2-4 part version number of non-negative int32 numbers,
+                // but mscorlib should be versioned correctly
+                var corlibVersion = new Version(corlibFileVersionString);
+                var corlib461Version = new Version(corlib461FileVersionString);
+                frameworkBasedSubdir = corlibVersion < corlib461Version ? "net45" : "net461";
+            }
+
+            string tracerHomeDirectory = ReadEnvironmentVariable("DD_DOTNET_TRACER_HOME") ?? string.Empty;
+            s_managedProfilerDirectory = Path.Combine(tracerHomeDirectory, frameworkBasedSubdir);
+        }
+
+        private bool TryFindAssemblyInProfilerDirectory(AssemblyName assemblyName, out string fullPath)
+        {
+            fullPath = Path.Combine(s_managedProfilerDirectory, $"{assemblyName.Name}.dll");
+            return File.Exists(fullPath);
+        }
+
+        private bool ShouldLoadAssemblyFromProfilerDirectory(AssemblyName assemblyName)
+        {
+            if (string.IsNullOrEmpty(assemblyName?.Name) || string.IsNullOrEmpty(assemblyName?.FullName))
+            {
+                return false;
+            }
+
+            // The AssemblyResolveEventHandler will handle all assemblies that have these prefixes:
+            // If the could not be found, we will look for them in the TRACER_HOME:
+
+            bool shouldHandle = (assemblyName.Name.StartsWith("Datadog.Trace", StringComparison.OrdinalIgnoreCase) == true)
+                    || (assemblyName.Name.StartsWith("Datadog.AutoInstrumentation", StringComparison.OrdinalIgnoreCase) == true);
```

Comment:
Can we put the assembly name list outside the method, because right now we have the SxS dictionary, this custom logic and also the list handled in the native side.

I prefer at least to have this and the SxS dictionary in the same part of the file or in another constants file.

Reply:
That makes sense. Have not made the change yet (wanna do some additional clean-up), but will do in the Updated in the _macrogreg/loader-with-modules-cctor-and-managed-loader_-branch.